### PR TITLE
Update to work with Catbox 10 (Hapi 17)

### LINKED ISSUE #76: Support catbox 10 (hapi v17)

### PR DESCRIPTION
**Pull request by EgoAleSum**
This is a first attempt. I'm still working on updating the tests (there's so many of them!) and plan to finish them tomorrow. Sharing this in the meanwhile.

Once complete, should fix #76

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -28,13 +28,9 @@ exports = module.exports = internals.Connection = function (options) {
...
 
-internals.Connection.prototype.start = function (callback) {
-
-    callback = Hoek.once(callback);
-
-    const self = this;
+internals.Connection.prototype.start = function () {
     if (this.client) {
-        return Hoek.nextTick(callback)();
+        return;
```
Comment (Marsup) [inline lib/index.js]: I would expect a consistent return value, so a `Promise.resolve()` here would make sense.

File: lib/index.js [hunk 2]
```diff
@@ -109,96 +104,104 @@ internals.Connection.prototype.validateSegmentName = function (name) {
...
-
-    if (!this.client) {
-        return callback(new Error('Connection not started'));
-    }
-
-    this.client.get(this.generateKey(key), (err, result) => {
-
-        if (err) {
-            return callback(err);
+internals.Connection.prototype.get = function (key) {
```
Comment (Marsup) [inline lib/index.js]: You should probably make this function async, this way you can just throw the 1st error, and use the promises that ioredis already exposes. Same goes for all the functions below.

File: package-lock.json [hunk 3]
```diff
@@ -0,0 +1,1704 @@
+{
```
Comment (Marsup) [inline package-lock.json]: We don't put package-locks in repos, please remove it and add it to gitignore.

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "catbox-redis",
   "description": "Redis adapter for catbox",
-  "version": "3.1.1",
+  "version": "4.0.0",
```
Comment (Marsup) [inline package.json]: Don't bump the version yourself, let the maintainer do it.

**Comment by EgoAleSum** [comment]
Ok, thanks for the comments. I've made all changes you requested, and thanks for pointing out that ioredis had support for promises, code looks much better now :)

One thing I don't really agree with you (though I've done it anyways) is the removal of the package-lock.json file. As per the [official documentation](https://docs.npmjs.com/files/package-lock.json): "This file is intended to be committed into source repositories [...]"

**Comment by EgoAleSum** [comment]
Ok, it's done @Marsup

**Comment by EgoAleSum** [comment]
(Also, added a test to ensure #59 works)

**Comment by Marsup** [comment]
I don't think you understood what I was saying. Making a function async doesn't mean putting a comment on it, you have to use the `async` keyword. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

**Comment by EgoAleSum** [comment]
@marsup I am aware of what an async function is :) the functions are now all asynchronous because they return a promise (directly from ioredis whenever possible).

They can be used with ES2016 async/await (in fact, most tests do use that feature).

**Comment by Marsup** [comment]
Then we disagree on the style you used to achieve it.

**Comment by EgoAleSum** [comment]
The start method can not be written using async/await (or, at least, it would look nastier). I stuck to promises for coherence with that, and for coherence with the fact that ioredis returns promises already. I don’t believe the Hapi style guide had a prescription on what to use, so I went with one.

From an end-user standpoint, they’re completely equivalent: the APIs can be consumed using promises or async/await interchangeably.

I honestly spent my entire Sunday on this, i’m not really looking forward rewriting most of the code again... you do have write access to my fork I believe, however.

**Comment by EgoAleSum** [comment]
Sorry for sounding a bit harsh in the comment last night. Fact is: I’m not a developer in my day-time job, and I’m doing this for a side project of mine. Because of time, I can only work on it on certain weekends.

If you want me to make the change, I can do that, but right now looks like it will have to wait until the last weekend of November.

I don’t want to delay this PR until then, so if anyone is able to change the code, s/he is welcome to do so. Tests would remain the same, so it shouldn’t be as much work as it was for me.

**Comment by pkaske** [comment]
I opened a [pr](https://github.com/EgoAleSum/catbox-redis/pull/1) on your fork to help you out, @EgoAleSum

**Comment by EgoAleSum** [comment]
@pkaske thanks for doing that! I've looked at it and it looks good, minus a couple small edits

**Comment by EgoAleSum** [comment]
Ok I managed to carve out a couple hours to make the changes myself to @pkaske 's PR. Thanks for the input!

I've also updated the tests, and linted the code. Had to revert the change to go from "Redis.createClient" to "new Redis" because that did not work with sentinels (not sure why, and not feeling like debugging it right now :) )

**Comment by EgoAleSum** [comment]
@Marsup does the code look good to you now?

**Comment by EgoAleSum** [comment]
@Marsup can you merge?

**Comment by Marsup** [comment]
This looks mostly good to me, nothing that can't be fixed post-merge, but @LoicMahieu is the maintainer of this module, that is if he still wants it, so he's the one who should give the go ahead, I was merely giving a hand in the review process.

**Comment by LoicMahieu** [comment]
Hi @EgoAleSum @Marsup 

The changes looks good to me. I will release it as v4.0.0

Thanks you both for your work.

**Comment by LoicMahieu** [comment]
+ catbox-redis@4.0.0

**Pull request merged**